### PR TITLE
fixes #27: Update DbTableGateway typehints to use AbstractTableGateway

### LINKED ISSUE #27: Update DbTableGateway typehints
Per [this thread on zf-apigility](https://github.com/zfcampus/zf-apigility/pull/169#discussion_r72506560), the `DbTableGateway` constructor currently requires a concrete `Zend\Db\TableGateway\TableGateway` instance, but should likely accept `TableGatewayInterface` instead. This would be a type loosening, which does not pose a BC break, but which would simplify mocking, and allow for alternate implementations when desired.

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes #27

### CODE REVIEW

File: src/Adapter/DbTableGateway.php [hunk 1]
```diff
@@ -11,20 +11,20 @@
...
      *
-     * @param TableGateway                      $tableGateway
+     * @param TableGatewayInterface             $tableGateway
      * @param null|Where|\Closure|string|array  $where
      * @param null|string|array                 $order
      * @param null|string|array                 $group
      * @param null|Having|\Closure|string|array $having
      */
-    public function __construct(TableGateway $tableGateway, $where = null, $order = null, $group = null, $having = null)
+    public function __construct(TableGatewayInterface $tableGateway, $where = null, $order = null, $group = null, $having = null)
```
Comment (froschdesign) [inline src/Adapter/DbTableGateway.php]: [`TableGatewayInterface`](https://github.com/zendframework/zend-db/blob/master/src/TableGateway/TableGatewayInterface.php) doesn't define the methods [`getSql`](https://github.com/zendframework/zend-paginator/blob/master/src/Adapter/DbTableGateway.php#L29) and [`getResultSetPrototype`](https://github.com/zendframework/zend-paginator/blob/master/src/Adapter/DbTableGateway.php#L44).

Comment (samsonasik) [reply src/Adapter/DbTableGateway.php]: any suggestion? or previously already correct?

Comment (samsonasik) [reply src/Adapter/DbTableGateway.php]: I think `AbstractTableGateway` is the correct way

**Comment by samsonasik** [comment]
@froschdesign I've changed to `AbstractTableGateway` which has methods `getSql()` and `getResultSetPrototype()`.  Travis failure is not related with this change.

**Comment by froschdesign** [comment]
@samsonasik 
`AbstractTableGateway` is correct and much better, but it doesn't fix the entire problem: the typehint should be an interface.

**Review by froschdesign — APPROVED** [review]

**Comment by vaclavvanik** [comment]
Typehinting against `TableGatewayInterface` is not possible without BC break. `AbstractTableGateway` is imho much better as @froschdesign noted. But how to fix typehinting against interface? Create new `DbTableGatewayInterface` class or wait for next major version?

**Comment by froschdesign** [comment]
@vaclavvanik 
> …wait for next major version?

We will wait and in the meantime `AbstractTableGateway` is good choice.

File: src/Adapter/DbTableGateway.php [hunk 2]
```diff
@@ -11,20 +11,20 @@
...
      *
-     * @param TableGateway                      $tableGateway
+     * @param AbstractTableGateway              $tableGateway
      * @param null|Where|\Closure|string|array  $where
      * @param null|string|array                 $order
      * @param null|string|array                 $group
      * @param null|Having|\Closure|string|array $having
      */
-    public function __construct(TableGateway $tableGateway, $where = null, $order = null, $group = null, $having = null)
+    public function __construct(AbstractTableGateway $tableGateway, $where = null, $order = null, $group = null, $having = null)
```
Comment (weierophinney) [inline src/Adapter/DbTableGateway.php:27]: I was going to ask why you were type-hinting on `AbstractTableGateway` and not `TableGatewayInterface`, and then noticed that we call the `getSql()` method, which is only available in the abstract implementation...

**Comment by weierophinney** [comment]
> But how to fix typehinting against interface? Create new DbTableGatewayInterface class or wait for next major version?

If you want to add such an interface to zend-db that contains the methods used here and in other packages, we can then adopt that later. Since implementations would likely implement that interface, this would be a scope widening change, which is backwards compatible.

In the meantime, adopting the abstract implementation will still work with any `TableGateway` extensions, and provides the ability to use classes that only extend `AbstractTableGateway`.

**Comment by weierophinney** [comment]
Thanks, @samsonasik!

**Pull request merged**